include/deal.II/lac/full_matrix.h

```diff
@@ -68,6 +70,12 @@ template <typename number>
 class FullMatrix : public Table<2, number>
 {
 public:
+  // The assertion in full_matrix.templates.h for whether or not a number is
+  // finite is not compatible for AD number types.
+  static_assert(
+    !Differentiation::AD::is_ad_number<number>::value,
+    "The FullMatrix class does not support auto-differentiable numbers.");
```

Comment:
This isn't wrong, but it drags a *lot* of code into this header file. Not pushing for this, just asking: Is there an alternative *positive* condition we could check? This would also rule out other possible problematic uses, not just AD types.

Reply:
I suppose we could check that all numbers are floating point numbers:
```c++
static_assert(std::is_arithmetic<number>::value,...)
```
Do we expect these two classes to support complex numbers as well?

Reply:
Good find. `std::is_arithmetic` won't quite work for `std::complex` (which we want to support): `is_arithmetic` equals `is_integral` or `is_floating_point`, where the latter does not support `std::complex`, see `https://en.cppreference.com/w/cpp/types/is_arithmetic 

Could we use `numbers::NumberTraits` somehow, or at least expand it to what we want?